// pops/pops.applet.js
/* pops.applet : Used to load and process son applets */

if(2) {//- vars.
	var X=exports, G=global
	,  modName='pops.widget'

	,	$fs=require('fs')
		,	$fs_readFile=$fs.readFile
		,	$fs_readFileSync=$fs.readFileSync
		,	$fs_stat=$fs.stat
	,	$path=require('path')
		,	$path_dirname=$path.dirname
		,	$path_join=$path.join

	,  pc=require('pops/pops.core')
		,	ArrClone=Array.Clone
		,	ArrCopyTo=Array.CopyTo
		,	Class=pc.Class
		,	cout=pc.cout
		,	ObjCopyTo=Object.CopyTo
		,	ObjClone=Object.Clone
		,	Property=pc.Property
	,	pcss=require('./pops.css')
		,	pcss_cssReset=pcss.cssReset
		,	pcss_Code=pcss.Code
		,	pcss_LoadFile=pcss.LoadFile
		,	pcss_LoadFileSync=pcss.LoadFileSync
	,  pfsl=require('./fs/pops.fs.local')
		,	pfsl_FindFile=pfsl_FindFile
		,	pfsl_FindFileSync=pfsl_FindFileSync
	,  phtml=require('pops/pops.html')
		,	phtml_Element=phtml.Element
		,	GuiItem=phtml.GuiItem
	,  phttp=require('pops/pops.http')

	,	autoId=0
	,	LoadFile, LoadFileSync
	,	appletManager
	,	appletBase
	,	BuildAppletGui
	;
};

LoadFileSync=X.LoadFileSync=function(filnam, ops) {
	//out('pops.widget.LoadFileSync');
	var fil1, fil2, fromOps, i, jj, k, k2, kk, l, z, z2, zz
	,	o=ops||{}
	,	rv={ $$isPopsApplet: 2 }
	,	_fs=o.fileSystem||pfsl
	;

	if(fil1=_fs.FindFileSync(filnam, ops)) {
		z=_fs.ReadJsonFileSync(fil1.path);
		//out('fil1.path='+fil1.path+'  -  fil1.dir='+fil1.dir);
		fromOps=ObjCopyTo({}, [ops||{}, { fromDir: fil1.dir }]);
		rv.infoPath=zz=fil1.path; rv.infoDir=fil1.dir;
		rv.name=z.name||'<not named>';
		rv.version=z.version||'0';
		rv.elem=z.elem;
		rv.props=z.props;
		if(k=rv.class=z.class)
			rv.classStr=k.toString().ReplaceAll(',', ' ');

		if(!z.client) throw(new Error('Applet must have client script'));
		else {
			if(fil2=_fs.FindFileSync(z.client, fromOps)) {
				rv.clientPath=fil2.path; rv.clientDir=fil2.dir;
				rv.clientString=_fs.readFileSync(fil2.path).toString();
				//out('rv.clientString <\n'+rv.clientString+'\n>\n');
				if(k=z.css) {
					if(!(k instanceof Array)) k=[k];
					zz=rv.css=[]; kk=0;
					for(i=0, l=k.length; i<l; i++) {
						z2=jj=k[i];
						if(typeof z2=='string') z2={ filename: jj };
						if(z2 && z2.filename) {
							z2.$$isCssFromFile=2;
							if(kk=_fs.FindFileSync(z2.filename, fromOps)) {
								if(z2.reload) z2.filename=kk.path; 
								else z2=pcss_LoadFileSync(kk.path);
								//else z2=pcss_LoadFileSync(kk.path);
							}
							else throw(new Error('file not found "'+z2.filename+'"'));
						};
						if(z2) zz.Push(z2);
					};
					
				};
				if(k=z.server) {
					if(fil2=_fs.FindFileSync(k, fromOps)) {
						rv.serverPath=fil2.path; rv.serverDir=fil2.dir;
						rv.server=_fs.require(fil2.path);
					}
					else
						throw(new Error('Can\'t find server file : '+k));
				}
			}
			else
				throw(new Error('Can\'t find client file : '+z.client));
		};

	}
	else
		throw(new Error('Can\'t find file : '+filnam));

	return rv;
};


if(2) {//- appletManager
	if(2) {//- private
		var
			amReset=function(cb) {
				if(cb) cb(0, this);
			}
		,	amResetSync=function() {
				var i, l, nam, nm, z, z2, zz
				,	o=this.OP||{}
				,	sp=this.$$searchPaths=o.searchPaths||[]
				,	applt=this.$$applets={}
				,	oApplt=o.applets||[]
				,	als=this.$$alias={}
				,	oAls=o.alias||{}
				,	_fs=o.fileSystem
				;

				if(oWgt.$$isPopsWidgetManager) wgt=oWgt.$$widgets||{}
				else
					for(i=0, l=oApplt.length; i<l; i++) {
						z2=0;
						z=oApplt[i];
						
						if(z.$$isPopsApplet) z2=z;
						else z2=LoadFileSync(z, {
							searchPaths: sp
						,	fileSystem: _fs||pfsl
						});
						
						if(z2) {
							if(!(nam=z2.name)) throw(new Error('name missing in applet info : '+z));
							else if(applt[nam]) throw(new Error('applet already loaded : '+z));
							else applt[nam]=z2;
						};
					};

				if(oAls.$$isPopsAppletManager) oAls=oAls.$$alias||{}
				for(nm in oAls) {
					z2=0;
					z=oAls[nm];
					//out('oAls['+nm+'].name = '+oAls[nm].name);
					if(applt[z]) {}
					else if(z.$$isPopsApplet) {
						applt[(nam=z.name)]=z;
						als[nm]=nam;
					}
					else {
						z2=LoadFileSync(z, {
							searchPaths: sp
						,	fileSystem: _fs||pfsl
						});
						if(!z2) throw(new Error('can\'t open file "'+z+'"'));
						else {
							if(!(nam=z2.name)) throw(new Error('name missing in applet info : '+z));
							else if(applt[nam]) throw(new Error('applet already loaded : '+z));
							else applt[nam]=z2;
							als[nm]=nam;
						};
					};
					
				};



			}
		;
	};
	appletManager=X.appletManager=Class('pops.applet:appletManager', {
		OPTIONS: {
			async: 0
		,	auto: 2
		,	alias: {}
		,	fileSystem: pfsl
		,	searchPaths: []
		,	applets: []
		}
	,	SHARED: {}
	,	INIT: function(ops, cb) {
			if(typeof ops=='function') { cb=ops; ops=0 };
			
			if(ops && ops.$$isPopsAppletManager) {
				this.alias=ObjClone(ops.alias);
				this.searchPaths=ArrClone(ops.searchPaths);
				this.applets=ObjClone(ops.applets);
			}
			else if(ops && ops.auto) this.Reset(ops, cb);
			else if(cb) cb(0, this);
		}
	,	FUNCTION: function(v) {
			//out('FUNCTION  -  v='+v);
			var z;
			if(z=this.$$alias[v]) v=z;
			return this.$$applets[v];
		}
	,	PUBLIC: {
			$$isPopsAppletManager: 2
		,	$$alias: 0
		,	$$searchPaths: 0
		,	$$applets: 0
	
		,	BuildAppletGui: function(ops, spacer, space, ender, cb) {
				return BuildAppletGui(
					ObjCopyTo({}, [ops, { appletManager: this } ])
				,	spacer, space, ender, cb
				);
			}
		,	Reset: function(ops, cb) {
				if(typeof ops=='function') { cb=ops; ops=0; };
				var o=this.SetOptions(ops||this.$options||{}).OP;
				if(o.async)
					amReset.apply(this, [cb]);
				else {
					amResetSync.apply(this);
					if(cb) cb(0, this);
				};
			}
		,	Start: function(ops, cb) { this.Reset(ops, cb); }
		}
	});
};

if(2) {//- BuildAppletGui
	BuildAppletGui=X.BuildAppletGui=function(ops, spacer, space, ender, cb) {
		var rv, css, e, el, iCode, id, k, kk, opts, props, z, z2, zz
		,	o=(ops)? ObjClone(ops) : {}
		,	BuildGui=o.BuildGui
		,	guiItm=o.guiItem
		,	appltMan=o.appletManager
		,	appltDat
		,	name=o.name
		,	aName
		;

		if(!guiItm) {//- error handler
			z=new Error('ops (for function "pops.applets.BuildAppletGui") must contain a "pops.html.GuiItem" instance named "guiItem"');
			if(cb) cb(z);
			else throw(z);
			return;
		};
		if(!BuildGui) {//- error handler
			z=new Error('ops (for function "pops.applets.BuildAppletGui") must contain "pops.spa.BuildGui" named "BuildGui"');
			if(cb) cb(z);
			else throw(z);
			return;
		};
		if(!appltMan) {//- error handler
			z=new Error('ops (for function "pops.applets.BuildAppletGui") must contain a "pops.widgets.widgetManager" instance named "widgetManager"');
			if(cb) cb(z);
			else throw(z);
			return;
		};
		if(!(appltDat=appltMan(name))) {//- err handler & var setter
			z=new Error('applet "'+name+'" not loaded. (for function "pops.applets.BuildAppletGui")');
			if(cb) cb(z);
			else throw(z);
			return;
		};		

		aName=appltDat.name;
		props=(z=guiItm.props)? ObjClone(z) : {};
		opts=props.options=(z=props.options)? ObjClone(z) : {};
		//out('props='+JSON.stringify(props));
		id=props.id||'AUTO_APPLET_ID_'+autoId++;
		if((z2=appltDat.css) && z2.length) css=ArrClone(z2);
		if(z=appltDat.classStr) props.class=z;
		
		if(z=appltDat.props) props=ObjCopyTo({}, [z, props]);
		el=o.outline=GuiItem(appltDat.elem||'div', id, props);
		if((z=appltDat.server) && z.SetupElem) {
			e={};
			z.SetupElem(e, el, props, opts);
		};

		if(!(z=el.props)) z=el.props={};
		z.id=id;
		if(opts=z.options) delete z.options;
		
		rv=BuildGui(o, spacer, space, ender);
		rv.applets[aName]=2;
		zz=rv.iCode||'';
		if(0) rv.iCode=(
			'$appltMan.GetSync("'+aName+'").Create('
		+		'$("'+id+'")'
		+	', '+JSON.stringify(opts||{ _0929_: 'RJM!' })
		+	');\n'
		+	zz
		);
		rv.iCode=(
			'z=new ($appltMan.GetSync("'+aName+'").applet)('
		+		'$("'+id+'")'
		+	', '+JSON.stringify(opts||{ _0929_: 'RJM!' })
		+	');\n'
		+	zz
		);
		k=rv.css=rv.css||[];
		if(kk=appltDat.css) k.push(kk);

		if(cb) cb(0, rv);
		return rv;
	}
};

if(2) {//- CreateApplet
	appletBase={
		$$isPopsAppletApplet: 2
	,	name: 'Applet must have unique name'
	,	outline: {}

	,	Init: Function.Empty //- args=(e, ops)
	,	SetupElem: Function.Empty //- args=(e, ops)
	}
	X.CreateApplet=function(tgtObj, nam, ops) {
		if(typeof nam=='object' && ops) { ops=nam; nam=0; };
		tgtObj.applet=ObjCopyTo({}, [
			appletBase
		,	ops
		,	{
				name: nam||ops.name
			}
		]);
	}
}
